doc/developer/design/20200520_exactly_once_kafka.md

```diff
@@ -1,28 +1,79 @@
+ - Read from its assigned partitions if any, and assign incoming records the appropriate timestamps. If an appropriate binding does not exist the operator needs to extend the timestamp binding for the current timestamp.
+- If the operator minted new timestamp bindings it needs to commit them back to worker-local state before exiting, so that those bindings can be shared with its peers.
+
+Whew. This is a complex (at least for me) protocol that involves some performance changes (now every time we need to timestamp a record we may need to binary search a list of timestamp bindings) to source ingestion code. The only alternative approach which made sense to me for this was to be a Kafka read replica, and write down `(data, timestamp)` for all incoming data, and have only one reader read the topic, and have every other reader read the written down state. This approach was scrapped because of the dependence on source persistence, which is another large project and currently a work in progress.
+
+Once we do the above, we also need a mechanism to compact the list of timestamp bindings in memory, as otherwise we risk running out of memory over time. We can do that rebinding all intervals assigned to a timestamp <= `compaction_frontier` to the `compaction_frontier`. At this stage, we also changed the dataflow worker to communicate source timestamp binding `upper` information back to the `Coordinator` so that the `Coordinator` can direct the dataflow worker on what frontier to compact up to based on the `logical_compaction_window`. After that, the `Coordinator` needs to also be modified to be aware of each source's `upper` and `since` frontiers and to disallow queries outside of the interval `[since, upper)` the way it does for indexes.
+
+2. We chose to persist timestamp binding information in a table in SQLite mostly for convenience because SQLite is already used as a storage layer for the catalog. Dataflow workers will send newly minted timestamp bindings to the `Coordinator`, which will persist them, and maintain a `durability_frontier` and send `durability_frontier` updates back to all the dataflow workers. Periodically, the `Coordinator` will also compact the persisted bindings that are at times less than or equal to the current `since` frontier.
+
+This approach is kind of like a "speculative execution" approach because updates can flow through the dataflow layer at times that have not yet been persisted to disk. We chose this approach rather than having the dataflow worker itself persist timestamps before downgrading capabilities because that puts the latency of persisting on the fast path to returning results, even when there are no sinks, and SQLite doesn't love multithreaded writes (this is a less good reason). The downside of this approach is that now we need to communicate a `durability_frontier` in addition to the standard `input_frontier` for operators that care about persistence.
```

Comment:
Correct me if wrong, but it feels like the choice is much more about SQLite. Workers could certainly release data without blocking on persisting bindings, and imo that seems like a great design. The issue seems to be that it is a giant pain to use SQLite from multiple threads.

Reply:
Mentioning mostly because this is a design doc (I think) and we want to write down the intended design, and I think "involve coordinator in all persistence work" isn't something we want to bake in as part of the design.

Reply:
you're definitely not wrong